workflow/README.md

```diff
@@ -0,0 +1,316 @@
+| `exception`    | `Task.Wasm`                 |                                           | No       | `null`  |
+| `signature`    | Varsig                      | Signature of serialized fields            | Yes      |         |
+
+``` ipldsch
+FIXME
+
+type Workflow struct {
+  ver String
+  req DID
+  nnc String 
+  vfy Verification 
+  meta &{String:String} implicit {}
+  par &Task optional nullable
+  dfl Config implicit {}
+  tsks {String: Task}
+  exc &Wasm optional nullable
+  sig
+}
+
+type Verification union {
+  | OptimisticVerification
+  | "snark"
+} representation keyed
+
+type OptimisticVerification struct {
+  optimistic Integer
+}
+```
+
+## 2.1 Fields
+
+## 2.1.1 Type
+
+The `type` field MUST be set to `ipvm/workflow`. This field together with the `version` field indicates the expected fields and minimal semantics for the workflow.
+
+## 2.1.2 Version
+
+The `version` field MUST be set to `0.1.0`. This field together with the `type` field indicates the expected fields and minimal semantics for the workflow.
+
+## 2.1.3 Requestor
+
+The `requestor` field MUST be set to the DID of the agent requesting the workflow. The Rquestor is the only identified agent in a Workflow.
+
+The `signature` field MUST validate with a public key associated with the Requestor's DID.
+
+## 2.1.4 Nonce
+
+The `nonce` field MUST be a one-time-use random string. At least 12 random bytes encoded as base64 is RECOMMENDED.
+
+## 2.1.5 Parent
+
+The OPTIONAL `parent` field contains the CID of the IPVM Task that initiated it (if any).
+
+## 2.1.6 Meta
+
+The `meta` field contains a user-definable JSON object. This is useful for including things like tags, comments, and so on.
+
+<!-- FIXME is this an attack vector? We should force this to be a CID I guess? -->
+
+## 2.1.7 Defaults
+
+The global `defaults` object (FIXME section X.Y) sets the configuration for the workflow itself, and defaults for tasks.
+ 
+## 2.1.8 Run
+
+The `run` field contains all of the IPVM Tasks set to run in this Workflow, each labelled by a human-readable key.
+
+See the [Task](FIXME) section for more.
+
+## 2.1.9 Exception
+
+The `exception` field contains a Task with predefined inputs. See the [Exception Handling](FIXME) section for more.
+
+## 2.1.10 Signature
+
+The signature of the CID represented by the other fields.
+
+## 2.2 Example
+
+Here is a simple example:
+
+```
+┌─────────┐   ┌─────────┐
+│         │   │         │
+│  left   │   │  right  │
+│         │   │         │
+└────────┬┘   └─┬───────┘
+         │      │
+         │      │
+       ┌─▼──────▼┐
+       │         │
+       │   end   │
+       │         │
+       └─────────┘
+```
+
+Here, two tasks (`left` and `right`) are used as input to a third task (`end`). This is fully configured in IPVM as:
+
+``` json
+{
+  "type": "ipvm/workflow",
+  "version": "0.1.0",
+  "requestor": "did:key:zAlice",
+  "nonce": "o3--8Gdu5",
+  "verification": {"optimistic": 2},
```

Comment:
There's a lot in here! And we are coming to the conclusion (like Koii) that verification can be quite task specific... and hence probably another user-definable object.

Reply:
Interesting — do you mean defining arbitrary user-defined verification types, or pushing this field into each task?

How does Koii do verification?

Reply:
My understanding is only from watching Al Morris' talk, but he seemed to suggest that it is up to the owner of the Task to define how it is verified. This opens up doing quite specific things for verification, e.g. he mentioned that deterministic work over data might be verified via hashing, but non-deterministic work like web scraping might be verified stochastically, in a domain-specific way.